### PR TITLE
Use invokable actions

### PR DESCRIPTION
**Pull request by Wharenn**
I am targeting this branch, because this is backward compatible.

## Changelog

```markdown
### Changed
- `Controller\AdminResettingController` is now deprecated in favor of `Action\{CheckEmail,Request,Reset,SendEmail}Action`
- `Controller\AdminSecurityController` is now deprecated in favor of `Action\{CheckLogin,Login,Logout}Action`
```

## To do
    
- [x] Deprecated `AdminResettingController`
- [x] Deprecated `AdminSecurityController`

## Subject

As discussed in #1052, this PR deprecates `AdminResettingController` in favor of new Actions. Changes are based on the work done on this PR : https://github.com/sonata-project/SonataAdminBundle/pull/5120

This may be a first run, so comments are welcomed :)
~~I think we should also include the deprecation of `AdminSecurityController` to be consistent~~ (done)

### CODE REVIEW

File: src/Action/CheckEmailAction.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+
+final class CheckEmailAction extends Controller
+{
+    /**
+     * @var Pool
+     */
+    protected $adminPool;
```
Comment (core23) [inline src/Action/CheckEmailAction.php]: private (same for the others)

File: src/Action/ResetAction.php [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        $this->authorizationChecker = $authorizationChecker;
+        $this->router = $router;
+        $this->adminPool = $adminPool;
+        $this->templateRegistry = $templateRegistry;
+        $this->formFactory = $formFactory;
+        $this->userManager = $userManager;
+        $this->loginManager = $loginManager;
+        $this->translator = $translator;
+        $this->resetTtl = $resetTtl;
+        $this->firewallName = $firewallName;
```
Comment (core23) [inline src/Action/ResetAction.php]: You need `$this->logger = new NullLogger()`

File: src/Controller/AdminResettingController.php [hunk 3]
```diff
@@ -30,181 +35,47 @@ class AdminResettingController extends Controller
...
+        return $requestAction($this->getCurrentRequest());
     }
 
     /**
-     * @param Request $request
-     *
      * @return Response
      */
-    public function sendEmailAction(Request $request)
+    public function sendEmailAction()
```
Comment (core23) [inline src/Controller/AdminResettingController.php]: This is a BC break

**Comment by Hanmac** [comment]
Symfony 4.2 did deprecate Controller class for AbstractController
https://symfony.com/blog/new-in-symfony-4-2-important-deprecations
maybe that can be done there too if needed?

or do you even need any base class at all?

File: src/Action/RequestAction.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+     * @var Pool
+     */
+    protected $adminPool;
+
+    /**
+     * @var TemplateRegistryInterface
+     */
+    protected $templateRegistry;
+
+    public function __construct(AuthorizationCheckerInterface $authorizationChecker, RouterInterface $router, Pool $adminPool, TemplateRegistryInterface $templateRegistry)
```
Comment (greg0ire) [inline src/Action/RequestAction.php]:
```suggestion
    public function __construct(
        AuthorizationCheckerInterface $authorizationChecker,
        UrlGeneratorInterface $urlGenerator,
        Pool $adminPool,
        TemplateRegistryInterface $templateRegistry
)
```

File: src/Action/ResetAction.php [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+            try {
+                $this->loginManager->logInUser($this->firewallName, $user, $response);
+                $user->setLastLogin(new \DateTime());
+            } catch (AccountStatusException $ex) {
+                // We simply do not authenticate users which do not pass the user
+                // checker (not enabled, expired, etc.).
+                if ($this->logger) {
+                    $this->getLogger()->warning(sprintf(
+                            'Unable to login user %d after password reset',
```
Comment (greg0ire) [inline src/Action/ResetAction.php]:
```suggestion
                        'Unable to login user %d after password reset',
```

File: src/Action/ResetAction.php [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+                            $user->getId())
```
Comment (greg0ire) [inline src/Action/ResetAction.php]:
```suggestion
                        $user->getId()
```

File: src/Action/ResetAction.php [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+                    );
```
Comment (greg0ire) [inline src/Action/ResetAction.php]:
```suggestion
                ), ['exception' => $ex]);
```

File: src/Action/RequestAction.php [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+ */
+
+namespace Sonata\UserBundle\Action;
+
+use Sonata\AdminBundle\Admin\Pool;
+use Sonata\AdminBundle\Templating\TemplateRegistryInterface;
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Routing\RouterInterface;
```
Comment (greg0ire) [inline src/Action/RequestAction.php]:
```suggestion
use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
```

File: src/Action/RequestAction.php [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+final class RequestAction extends Controller
+{
+    /**
+     * @var AuthorizationCheckerInterface
+     */
+    protected $authorizationChecker;
+
+    /**
+     * @var RouterInterface
```
Comment (greg0ire) [inline src/Action/RequestAction.php]:
```suggestion
     * @var UrlGeneratorInterface
```

File: src/Action/RequestAction.php [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+     */
+    protected $router;
```
Comment (greg0ire) [inline src/Action/RequestAction.php]:
```suggestion
    private $urlGenerator;
```

File: src/Action/RequestAction.php [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+    /**
+     * @var TemplateRegistryInterface
+     */
+    protected $templateRegistry;
+
+    public function __construct(AuthorizationCheckerInterface $authorizationChecker, RouterInterface $router, Pool $adminPool, TemplateRegistryInterface $templateRegistry)
+    {
+        $this->authorizationChecker = $authorizationChecker;
+        $this->router = $router;
```
Comment (greg0ire) [inline src/Action/RequestAction.php]:
```suggestion
        $this->urlGenerator = $urlGenerator;
```

File: src/Action/RequestAction.php [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        $this->adminPool = $adminPool;
+        $this->templateRegistry = $templateRegistry;
+    }
+
+    public function __invoke(Request $request)
```
Comment (greg0ire) [inline src/Action/RequestAction.php]:
```suggestion
    public function __invoke(Request $request): Response
```

File: src/Action/RequestAction.php [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\UserBundle\Action;
+
+use Sonata\AdminBundle\Admin\Pool;
+use Sonata\AdminBundle\Templating\TemplateRegistryInterface;
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
```
Comment (greg0ire) [inline src/Action/RequestAction.php]:
```suggestion
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
```

File: src/Action/ResetAction.php [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+        $this->templateRegistry = $templateRegistry;
+        $this->formFactory = $formFactory;
+        $this->userManager = $userManager;
+        $this->loginManager = $loginManager;
+        $this->translator = $translator;
+        $this->resetTtl = $resetTtl;
+        $this->firewallName = $firewallName;
+    }
+
+    public function __invoke(Request $request, $token)
```
Comment (greg0ire) [inline src/Action/ResetAction.php]:
```suggestion
    public function __invoke(Request $request, $token): Response
```

File: src/Action/ResetAction.php [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+
+use FOS\UserBundle\Form\Factory\FactoryInterface;
+use FOS\UserBundle\Model\UserManagerInterface;
+use FOS\UserBundle\Security\LoginManagerInterface;
+use Psr\Log\LoggerAwareTrait;
+use Sonata\AdminBundle\Admin\Pool;
+use Sonata\AdminBundle\Templating\TemplateRegistryInterface;
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
```
Comment (greg0ire) [inline src/Action/ResetAction.php]:
```suggestion
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
```

File: src/Action/CheckEmailAction.php [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace Sonata\UserBundle\Action;
+
+use Sonata\AdminBundle\Admin\Pool;
+use Sonata\AdminBundle\Templating\TemplateRegistryInterface;
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+
+final class CheckEmailAction extends Controller
```
Comment (greg0ire) [inline src/Action/CheckEmailAction.php]: Bonus points if you avoid extending `Controller` by injecting the templating engine

**Comment by Hanmac** [comment]
@greg0ire your opinion to the Controller depreciation or do we not care yet?

**Comment by OskarStark** [comment]
I think he means there is no need to extend anything

**Comment by greg0ire** [comment]
> I think he means there is no need to extend anything

Yes, but as I said, "for bonus points", meaning I will not block this already great PR with that request.

**Comment by Wharenn** [comment]
I made an update by adding the actions of the `AdminSecurityController` and also including removal of the extends (clean actions yeah!). I think the scope of this PR is now complete. 

@greg0ire @core23, if there is any new changes you see, I will try to update asap :)

File: src/Action/RequestAction.php [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+
+    public function __invoke(Request $request): Response
+    {
+        if ($this->authorizationChecker->isGranted('IS_AUTHENTICATED_FULLY')) {
+            return new RedirectResponse($this->urlGenerator->generate('sonata_admin_dashboard'));
+        }
+
+        return $this->templating->renderResponse('@SonataUser/Admin/Security/Resetting/request.html.twig', [
+            'base_template' => $this->templateRegistry->getTemplate('layout'),
+            'admin_pool' => $this->adminPool,
```
Comment (core23) [inline src/Action/RequestAction.php]: Not part of this PR, but can we make a global twig variable for `base_template` and `admin_pool`, as every admin template need this

File: src/Action/ResetAction.php [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+     * @var int
+     */
+    private $resetTtl;
+
+    /**
+     * @var string
+     */
+    private $firewallName;
+
+    public function __construct(
```
Comment (core23) [inline src/Action/ResetAction.php]: OMG! Do we really need that much deps? oO

**Comment by core23** [comment]
I like the PR 👍  but should we pick the chance to reduce the dependencies for the classes first? Later that would be a BC break.

File: src/Action/LoginAction.php [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+
+            return new RedirectResponse($url);
+        }
+
+        $csrfToken = null;
+        if ($this->csrfTokenManager) {
+            $csrfToken = $this->csrfTokenManager->getToken('authenticate')->getValue();
+        }
+
+        return $this->templating->renderResponse('@SonataUser/Admin/Security/login.html.twig', [
```
[the remark this answers is not in the document]
Comment (Wharenn) [reply src/Action/LoginAction.php]: It looks like the `templating` service is defined by FrameworkBundle which is a requirement. So I think we always fell to this definition in case of rendering:
https://github.com/symfony/framework-bundle/blob/master/Resources/config/templating_php.xml
https://github.com/symfony/framework-bundle/blob/4.1/DependencyInjection/FrameworkExtension.php#L823

File: src/Action/CheckLoginAction.php [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\UserBundle\Action;
+
+final class CheckLoginAction
```
[the remark this answers is not in the document]
Comment (Wharenn) [reply src/Action/CheckLoginAction.php]: We need the routing of these actions for the login/logout process. What do you propose? A single catch action throwing an exception?

**Comment by Wharenn** [comment]
Do you see any clue to reduce the amount of dependencies? I do not see any optional dependency here. Dependencies listing is the "downside" of container-less controllers/actions as we're not relying anymore on the underlying container `get()` method for common actions like translations, rendering or flash messages.

**Comment by core23** [comment]
> Do you see any clue to reduce the amount of dependencies? 

Maybe we can use Events (like https://github.com/FriendsOfSymfony/FOSUserBundle/blob/master/Controller/ChangePasswordController.php) for some stuff, e.g. Flash Messages and Mailing

> I do not see any optional dependency here. Dependencies listing is the "downside" of container-less controllers/actions as we're not relying anymore on the underlying container `get()` method for common actions like translations, rendering or flash messages.

I know that the old Controllers had way more intransparent dependencies, the new action style is a huge improvement and made all dependencies transparent 👍

**Comment by Wharenn** [comment]
This approach is interesting. As you proposed to remove dependencies on the pool and template registry, it could be really cool to get rid of the session/flash with events. Maybe we could do this in a next PR. I must confess this has been a long run since the #1052 initial proposal 😅

**Comment by Hanmac** [comment]
@core23 interesting, i wonder because UserBundle already require FOSUser, so in theory we can use its events too?

**Comment by greg0ire** [comment]
Keep in mind that if this is done in a subsequent PR, it will be a BC-break unless we do not create a new release. Otherwise we will have to keep the extra constructor arguments and deprecate them.

**Comment by OskarStark** [comment]
@core23 can you approve and if so, please merge :)
